File: ios/BowtieKit/Sources/BowtieKit/BowtieClient.swift
import Foundation

// MARK: - Request / error wire shapes

private struct LoginBody: Encodable {
    let username: String
    let password: String
}

private struct RefreshBody: Encodable {
    let refreshToken: String
}

private struct ChangePasswordBody: Encodable {
    let currentPassword: String
    let newPassword: String
}

private struct CreateSessionBody: Encodable {
    let channelId: Int64
    let caps: ClientCaps
}

private struct ErrorBody: Decodable {
    let error: String
}

private struct TunersBusyBody: Decodable {
    let error: String
    let sessions: [ActiveSessionSummary]
}

// MARK: - Client

/// Viewer-only HTTP client for the Bowtie API.
///
/// Auth: bearer access token in memory; refresh token in `SessionStore`.
/// Refresh is single-flight via in-actor `Task` coalescing; the new refresh
/// token is persisted before any 401-retry fires.
public actor BowtieClient {
    private let server: URL
    private let store: SessionStore
    private let urlSession: URLSession

    public private(set) var currentUser: User?
    private var accessToken: String?

    /// In-flight refresh shared by concurrent 401 waiters.
    private var refreshTask: Task<Void, Error>?

    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .iso8601
        return d
    }()

    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        // Default keys are camelCase — matches OpenAPI field names.
        return e
    }()

    private let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    public init(server: URL, store: SessionStore, urlSession: URLSession = .shared) {
        self.server = server
        self.store = store
        self.urlSession = urlSession
    }

    // MARK: - Viewer allowlist

    public func login(username: String, password: String) async throws -> User {
        let pair: TokenPair = try await send(
            path: "/api/v1/auth/login",
            method: "POST",
            body: LoginBody(username: username, password: password),
            authorize: false,
            retryOn401: false
        )
        applyTokens(pair)
        return pair.user
    }

    /// Rotate the stored refresh token into a new pair. Throws `.unauthorized` if absent/dead.
    public func bootstrapFromStoredToken() async throws -> User {
        guard store.loadRefreshToken() != nil else {
            throw BowtieError.unauthorized
        }
        try await performRefresh()
        guard let user = currentUser else {
            throw BowtieError.unauthorized
        }
        return user
    }

    /// Best-effort logout: always clears local session even if the network call fails.
    public func logout() async {
        let token = store.loadRefreshToken()
        if let token {
            _ = try? await sendRaw(
                path: "/api/v1/auth/logout",
                method: "POST",
                body: RefreshBody(refreshToken: token),
                authorize: false,
                retryOn401: false
            )
        }
        clearSession(keepServer: true)
    }

    public func changePassword(current: String, new: String) async throws {
        _ = try await sendRaw(
            path: "/api/v1/me/password",
            method: "POST",
            body: ChangePasswordBody(currentPassword: current, newPassword: new),
            authorize: true,
            retryOn401: true
        )
    }

    public func channels() async throws -> [Channel] {
        try await send(
            path: "/api/v1/channels",
            method: "GET",
            body: nil as EmptyBody?,
            authorize: true,
            retryOn401: true
        )
    }

    public func guide(start: Date, stop: Date) async throws -> [GuideChannel] {
        var components = URLComponents(
            url: ServerURL.resolve(path: "/api/v1/guide", against: server),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [
            URLQueryItem(name: "start", value: isoFormatter.string(from: start)),
            URLQueryItem(name: "stop", value: isoFormatter.string(from: stop)),
        ]
        guard let url = components.url else {
            throw BowtieError.invalidServerURL
        }
        return try await sendURL(
            url: url,
            method: "GET",
            body: nil as EmptyBody?,
            authorize: true,
            retryOn401: true
        )
    }

    public func createSession(channelId: Int64, caps: ClientCaps) async throws -> CreatedSession {
        try await send(
            path: "/api/v1/sessions",
            method: "POST",
            body: CreateSessionBody(channelId: channelId, caps: caps),
            authorize: true,
            retryOn401: true
        )
    }

    /// Best-effort DELETE; errors are swallowed.
    public func deleteSession(viewerId: String) async {
        let encoded = viewerId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? viewerId
        _ = try? await sendRaw(
            path: "/api/v1/sessions/\(encoded)",
            method: "DELETE",
            body: nil as EmptyBody?,
            authorize: true,
            retryOn401: true
        )
    }

    public func me() async throws -> User {
        let user: User = try await send(
            path: "/api/v1/me",
            method: "GET",
            body: nil as EmptyBody?,
            authorize: true,
            retryOn401: true
        )
        currentUser = user
        return user
    }

    // MARK: - Test hooks (@testable)

    /// Seeds the in-memory access token without a network round-trip.
    func setAccessTokenForTesting(_ token: String?) {
        accessToken = token
    }

    /// Builds a URLRequest applying the same Authorization policy as live calls.
    func makeURLRequestForTesting(
        path: String,
        method: String,
        authorize: Bool
    ) -> URLRequest {
        let url = ServerURL.resolve(path: path, against: server)
        return makeRequest(url: url, method: method, bodyData: nil, authorize: authorize)
    }

    // MARK: - Token / session state

    private func applyTokens(_ pair: TokenPair) {
        accessToken = pair.accessToken
        currentUser = pair.user
        // Persist new refresh BEFORE any 401-retry observes it.
        store.save(server: server, refreshToken: pair.refreshToken)
    }

    private func clearSession(keepServer: Bool) {
        accessToken = nil
        currentUser = nil
        if keepServer {
            store.save(server: store.loadServer() ?? server, refreshToken: nil)
        } else {
            store.save(server: nil, refreshToken: nil)
        }
    }

    // MARK: - Single-flight refresh

    private func refreshSingleFlight() async throws {
        if let refreshTask {
            try await refreshTask.value
            return
        }
        let task = Task {
            try await self.performRefresh()
        }
        refreshTask = task
        do {
            try await task.value
            refreshTask = nil
        } catch {
            refreshTask = nil
            throw error
        }
    }

    private func performRefresh() async throws {
        guard let refreshToken = store.loadRefreshToken() else {
            clearSession(keepServer: true)
            throw BowtieError.unauthorized
        }

        do {
            let pair: TokenPair = try await send(
                path: "/api/v1/auth/refresh",
                method: "POST",
                body: RefreshBody(refreshToken: refreshToken),
                authorize: false,
                retryOn401: false
            )
            applyTokens(pair)
        } catch let error as BowtieError {
            if case .unauthorized = error {
                clearSession(keepServer: true)
            }
            throw error
        }
    }

    // MARK: - HTTP

    private struct EmptyBody: Encodable {}

    private func send<B: Encodable, T: Decodable>(
        path: String,
        method: String,
        body: B?,
        authorize: Bool,
        retryOn401: Bool
    ) async throws -> T {
        let data = try await sendRaw(
            path: path,
            method: method,
            body: body,
            authorize: authorize,
            retryOn401: retryOn401
        )
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw BowtieError.network("decode failed: \(error.localizedDescription)")
        }
    }

    private func sendURL<B: Encodable, T: Decodable>(
        url: URL,
        method: String,
        body: B?,
        authorize: Bool,
        retryOn401: Bool
    ) async throws -> T {
        let data = try await sendRawURL(
            url: url,
            method: method,
            body: body,
            authorize: authorize,
            retryOn401: retryOn401
        )
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw BowtieError.network("decode failed: \(error.localizedDescription)")
        }
    }

    private func sendRaw<B: Encodable>(
        path: String,
        method: String,
        body: B?,
        authorize: Bool,
        retryOn401: Bool
    ) async throws -> Data {
        let url = ServerURL.resolve(path: path, against: server)
        return try await sendRawURL(
            url: url,
            method: method,
            body: body,
            authorize: authorize,
            retryOn401: retryOn401
        )
    }

    private func sendRawURL<B: Encodable>(
        url: URL,
        method: String,
        body: B?,
        authorize: Bool,
        retryOn401: Bool
    ) async throws -> Data {
        let bodyData: Data?
        if let body {
            bodyData = try encoder.encode(body)
        } else {
            bodyData = nil
        }

        let request = makeRequest(url: url, method: method, bodyData: bodyData, authorize: authorize)
        let (data, response) = try await perform(request)

        if let http = response as? HTTPURLResponse, http.statusCode == 401, retryOn401 {
            try await refreshSingleFlight()
            let retry = makeRequest(url: url, method: method, bodyData: bodyData, authorize: authorize)
            let (data2, response2) = try await perform(retry)
            return try mapSuccess(data: data2, response: response2)
        }

        return try mapSuccess(data: data, response: response)
    }

    private func makeRequest(
        url: URL,
        method: String,
        bodyData: Data?,
        authorize: Bool
    ) -> URLRequest {
        var request = URLRequest(url: url)
        request.httpMethod = method
        if let bodyData {
            request.httpBody = bodyData
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        if authorize, let token = accessToken, shouldAttachBearer(to: url) {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    /// Never attach Authorization to media/stream URLs (HLS uses `?token=`).
    private func shouldAttachBearer(to url: URL) -> Bool {
        let path = url.path
        return !path.contains("/api/v1/stream/")
    }

    private func perform(_ request: URLRequest) async throws -> (Data, URLResponse) {
        do {
            return try await urlSession.data(for: request)
        } catch {
            throw BowtieError.network(error.localizedDescription)
        }
    }

    private func mapSuccess(data: Data, response: URLResponse) throws -> Data {
        guard let http = response as? HTTPURLResponse else {
            throw BowtieError.network("non-HTTP response")
        }
        let status = http.statusCode
        switch status {
        case 200...299:
            return data
        case 401:
            throw BowtieError.unauthorized
        case 404:
            throw BowtieError.notFound
        case 422:
            let message = (try? decoder.decode(ErrorBody.self, from: data))?.error ?? "negotiation failed"
            throw BowtieError.negotiationFailed(message)
        case 503:
            if let busy = try? decoder.decode(TunersBusyBody.self, from: data) {
                throw BowtieError.tunersBusy(busy.sessions)
            }
            let message = (try? decoder.decode(ErrorBody.self, from: data))?.error ?? "service unavailable"
            throw BowtieError.server(status: status, message: message)
        default:
            let message = (try? decoder.decode(ErrorBody.self, from: data))?.error
                ?? HTTPURLResponse.localizedString(forStatusCode: status)
            throw BowtieError.server(status: status, message: message)
        }
    }
}

File: android/core/src/main/kotlin/app/bowtie/core/BowtieClient.kt
package app.bowtie.core

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import okhttp3.HttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.time.Instant

/**
 * Viewer-allowlist HTTP client: login/refresh/logout/me/password/channels/guide/sessions.
 *
 * Access token is memory-only. Refresh is **single-flight** (Mutex-coalesced): concurrent
 * 401s share one `/auth/refresh`; the new refresh token is persisted **before** retries.
 *
 * No OkHttp auth interceptor — Media3 must use an unauthenticated data source;
 * stream auth is the playlist `?token=` query (Global Constraints).
 */
class BowtieClient(
    val server: HttpUrl,
    private val store: TokenStore,
    private val okHttp: OkHttpClient = OkHttpClient(),
) {
    private val _currentUser = MutableStateFlow<User?>(null)
    val currentUser: StateFlow<User?> = _currentUser.asStateFlow()

    @Volatile
    private var accessToken: String? = null

    private val refreshMutex = Mutex()

    suspend fun login(username: String, password: String): User = withContext(Dispatchers.IO) {
        val pair = postUnauthed(
            path = "/api/v1/auth/login",
            bodyJson = BowtieJson.encodeToString(
                LoginRequest(username = username, password = password),
            ),
        )
        applyTokens(pair)
        pair.user
    }

    /**
     * Rotate the stored refresh token into a live session.
     * Throws [BowtieError.Unauthorized] when absent or rotation fails.
     */
    suspend fun bootstrapFromStoredToken(): User = withContext(Dispatchers.IO) {
        val rt = store.loadRefreshToken() ?: throw BowtieError.Unauthorized
        try {
            performRefresh(rt)
        } catch (_: BowtieError.Unauthorized) {
            throw BowtieError.Unauthorized
        } catch (_: BowtieError) {
            clearSessionKeepServer()
            throw BowtieError.Unauthorized
        }
        _currentUser.value ?: throw BowtieError.Unauthorized
    }

    /** Best-effort logout; always clears in-memory session and stored refresh token. */
    suspend fun logout() = withContext(Dispatchers.IO) {
        val rt = store.loadRefreshToken()
        if (rt != null) {
            try {
                val request = Request.Builder()
                    .url(apiUrl("/api/v1/auth/logout"))
                    .post(jsonBody(BowtieJson.encodeToString(RefreshRequest(refreshToken = rt))))
                    .header("Content-Type", JSON_MEDIA)
                    .build()
                okHttp.newCall(request).execute().close()
            } catch (_: Exception) {
                // best-effort
            }
        }
        clearSessionKeepServer()
    }

    suspend fun changePassword(current: String, new: String): Unit = withContext(Dispatchers.IO) {
        authed(
            method = "POST",
            path = "/api/v1/me/password",
            bodyJson = BowtieJson.encodeToString(
                ChangePasswordRequest(currentPassword = current, newPassword = new),
            ),
        )
    }

    suspend fun channels(): List<Channel> = withContext(Dispatchers.IO) {
        val body = authed("GET", "/api/v1/channels")
        BowtieJson.decodeFromString(body)
    }

    suspend fun guide(start: Instant, stop: Instant): List<GuideChannel> =
        withContext(Dispatchers.IO) {
            val path = "/api/v1/guide?start=${start}&stop=${stop}"
            val body = authed("GET", path)
            BowtieJson.decodeFromString(body)
        }

    suspend fun createSession(channelId: Long, caps: ClientCaps): CreatedSession =
        withContext(Dispatchers.IO) {
            val body = authed(
                method = "POST",
                path = "/api/v1/sessions",
                bodyJson = BowtieJson.encodeToString(
                    CreateSessionRequest(channelId = channelId, caps = caps),
                ),
            )
            BowtieJson.decodeFromString(body)
        }

    /** Best-effort delete; swallows all errors. */
    suspend fun deleteSession(viewerId: String): Unit = withContext(Dispatchers.IO) {
        try {
            authed("DELETE", "/api/v1/sessions/$viewerId")
        } catch (_: Exception) {
            // swallow
        }
    }

    suspend fun me(): User = withContext(Dispatchers.IO) {
        val body = authed("GET", "/api/v1/me")
        val user = BowtieJson.decodeFromString<User>(body)
        _currentUser.value = user
        user
    }

    // ── Auth / single-flight refresh ────────────────────────────────────────

    private fun applyTokens(pair: TokenPair) {
        accessToken = pair.accessToken
        // Persist new refresh BEFORE any retry can fire.
        store.save(server.toString(), pair.refreshToken)
        _currentUser.value = pair.user
    }

    private fun clearSessionKeepServer() {
        accessToken = null
        _currentUser.value = null
        store.save(store.loadServer(), null)
    }

    /**
     * Coalesce concurrent refresh attempts onto one network call.
     *
     * [failedAccessToken] is the Bearer that got 401. Waiters that acquire the
     * mutex after a successful rotation see a different access token and skip
     * (server refresh tokens are single-use — a second rotate would sign out).
     */
    private suspend fun singleFlightRefresh(failedAccessToken: String?) {
        refreshMutex.withLock {
            // Another coroutine already rotated past the token that failed.
            if (accessToken != null && accessToken != failedAccessToken) {
                return
            }
            // A prior refresh already cleared the session.
            if (store.loadRefreshToken() == null) {
                throw BowtieError.Unauthorized
            }
            try {
                val rt = store.loadRefreshToken()
                    ?: throw BowtieError.Unauthorized
                performRefresh(rt)
            } catch (e: BowtieError.Unauthorized) {
                throw e
            } catch (_: BowtieError) {
                clearSessionKeepServer()
                throw BowtieError.Unauthorized
            } catch (_: Exception) {
                clearSessionKeepServer()
                throw BowtieError.Unauthorized
            }
        }
    }

    private fun performRefresh(refreshToken: String) {
        val request = Request.Builder()
            .url(apiUrl("/api/v1/auth/refresh"))
            .post(jsonBody(BowtieJson.encodeToString(RefreshRequest(refreshToken = refreshToken))))
            .header("Content-Type", JSON_MEDIA)
            .build()
        try {
            okHttp.newCall(request).execute().use { response ->
                val body = response.body?.string().orEmpty()
                if (!response.isSuccessful) {
                    clearSessionKeepServer()
                    throw BowtieError.Unauthorized
                }
                val pair = BowtieJson.decodeFromString<TokenPair>(body)
                // Persist before any waiter retries.
                applyTokens(pair)
            }
        } catch (e: BowtieError) {
            throw e
        } catch (e: Exception) {
            throw BowtieError.Network(e)
        }
    }

    // ── HTTP helpers ────────────────────────────────────────────────────────

    private fun postUnauthed(path: String, bodyJson: String): TokenPair {
        val request = Request.Builder()
            .url(apiUrl(path))
            .post(jsonBody(bodyJson))
            .header("Content-Type", JSON_MEDIA)
            .build()
        try {
            okHttp.newCall(request).execute().use { response ->
                val body = response.body?.string().orEmpty()
                if (!response.isSuccessful) {
                    throw mapHttpError(response.code, body)
                }
                return BowtieJson.decodeFromString(body)
            }
        } catch (e: BowtieError) {
            throw e
        } catch (e: Exception) {
            throw BowtieError.Network(e)
        }
    }

    /**
     * Authenticated request. On 401: single-flight refresh then one retry.
     * Never attaches Bearer to `/api/v1/stream/` paths.
     */
    private suspend fun authed(
        method: String,
        path: String,
        bodyJson: String? = null,
        retryOn401: Boolean = true,
    ): String {
        val attachAuth = !isStreamPath(path)

        fun build(token: String?): Request {
            val b = Request.Builder().url(apiUrl(path))
            when (method) {
                "GET" -> b.get()
                "DELETE" -> b.delete()
                "POST" -> {
                    val rb = (bodyJson ?: "").toRequestBody(JSON_MEDIA_TYPE)
                    b.post(rb)
                    b.header("Content-Type", JSON_MEDIA)
                }
                else -> error("unsupported method $method")
            }
            if (attachAuth && token != null) {
                b.header("Authorization", "Bearer $token")
            }
            return b.build()
        }

        try {
            // Snapshot the token actually sent — concurrent refresh may rotate
            // accessToken before we process a late 401 for the old token.
            val tokenUsed = accessToken
            val first = okHttp.newCall(build(tokenUsed)).execute()
            try {
                if (first.code == 401 && retryOn401 && attachAuth) {
                    first.close()
                    singleFlightRefresh(failedAccessToken = tokenUsed)
                    okHttp.newCall(build(accessToken)).execute().use { retry ->
                        if (retry.code == 401) {
                            clearSessionKeepServer()
                            throw BowtieError.Unauthorized
                        }
                        return handleBody(retry)
                    }
                }
                return handleBody(first)
            } finally {
                first.close()
            }
        } catch (e: BowtieError) {
            throw e
        } catch (e: Exception) {
            throw BowtieError.Network(e)
        }
    }

    private fun handleBody(response: Response): String {
        val body = response.body?.string().orEmpty()
        if (response.isSuccessful) {
            return body
        }
        throw mapHttpError(response.code, body)
    }

    private fun mapHttpError(code: Int, body: String): BowtieError {
        return when (code) {
            401 -> BowtieError.Unauthorized
            404 -> BowtieError.NotFound
            422 -> BowtieError.NegotiationFailed(
                extractErrorMessage(body) ?: "negotiation failed",
            )
            503 -> {
                try {
                    val payload = BowtieJson.decodeFromString<TunersBusyPayload>(body)
                    BowtieError.TunersBusy(payload.sessions)
                } catch (_: Exception) {
                    BowtieError.Server(503, extractErrorMessage(body) ?: body)
                }
            }
            else -> BowtieError.Server(
                code,
                extractErrorMessage(body) ?: body.ifEmpty { "HTTP $code" },
            )
        }
    }

    private fun extractErrorMessage(body: String): String? {
        if (body.isBlank()) return null
        return try {
            BowtieJson.decodeFromString<ErrorBody>(body).error
        } catch (_: Exception) {
            null
        }
    }

    private fun apiUrl(path: String): HttpUrl {
        val qIndex = path.indexOf('?')
        return if (qIndex < 0) {
            ServerUrl.resolve(path, server)
        } else {
            val pathOnly = path.substring(0, qIndex)
            val query = path.substring(qIndex + 1)
            val base = ServerUrl.resolve(pathOnly, server)
            base.newBuilder().encodedQuery(query).build()
        }
    }

    private fun jsonBody(json: String) = json.toRequestBody(JSON_MEDIA_TYPE)

    companion object {
        private const val JSON_MEDIA = "application/json"
        private val JSON_MEDIA_TYPE = JSON_MEDIA.toMediaType()

        fun isStreamPath(path: String): Boolean =
            path.contains("/api/v1/stream/")
    }
}

// ── Wire request / error envelopes (OpenAPI field names) ────────────────────

@Serializable
private data class LoginRequest(val username: String, val password: String)

@Serializable
private data class RefreshRequest(val refreshToken: String)

@Serializable
private data class ChangePasswordRequest(
    val currentPassword: String,
    val newPassword: String,
)

@Serializable
private data class CreateSessionRequest(
    val channelId: Long,
    val caps: ClientCaps,
)

@Serializable
private data class ErrorBody(val error: String)

@Serializable
private data class TunersBusyPayload(
    val error: String,
    val sessions: List<ActiveSessionSummary> = emptyList(),
)
